### PR TITLE
Recruit with pending conditions button on manage

### PR DESCRIPTION
**Pull request by stevehook**
## Context
It would be beneficial to SCITT providers if we allowed them to move a candidate who was only pending SKE conditions to ‘Recruited’ as this would automatically import their record into Register saving them from having to manually create it. This is particularly useful for providers (and DfE) who are busy getting their ITT data ready for the ITT census deadline in October.

Implementing this will move a candidate to ‘Recruited’ status, meaning that they will be included in the next import from Apply to Register (a job that currently runs at 4am every day).

## Changes proposed in this pull request

- [x] A new _Recruit candidate with pending conditions_ button on the offer page:
image

- [x] The new button above leads to a new _Do you want to recruit the candidate with pending conditions?_ page.
image

- [x] If the use confirms the action the candidate is moved on the the `recruited` state. We display a flash message.

### To be covered in a follow up PR
- We need to display a _SKE pending conditions_ tag next to _Recruited_ tag wherever we display it.
image

## Guidance to review
- This PR _kind of_ introduces a new _Recruited with pending SKE conditions_ 'virtual status'. However we are modelling such applications using the existing _Recruited_ status - there is no new status value.
- For now we've avoided making any DB schema changes (to add a dedicated `recruited_with_pending_ske_conditions` flag). Instead we are going to work this out using existing data (the offer conditions and their status). If performance is an issue we may revisit this.
- Are the tests sufficient?

## Link to Trello card

[](https://trello.com/c/0V9VGfjL/6161-recruit-with-pending-conditions-button-on-manage)

## Things to check

- [x] If the code removes any existing feature flags, a data migration has also been added to delete the entry from the database
- [x] This code does not rely on migrations in the same Pull Request
- [x] If this code includes a migration adding or changing columns, it also backfills existing records for consistency
- [x] If this code adds a column to the DB, decide whether it needs to be in analytics yml file or analytics blocklist
- [x] API release notes have been updated if necessary
- [x] If it adds a significant user-facing change, is it documented in the [CHANGELOG](CHANGELOG.md)?
- [x] Required environment variables have been updated [added to the Azure KeyVault](/docs/environment-variables.md#deploy-pipeline)

### CODE REVIEW

File: app/forms/provider_interface/recruit_with_pending_conditions_form.rb [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+    validates :confirmation, presence: true
+    validates :confirmation, inclusion: { in: %w[yes no] }
+
+    def save
+      return false unless valid?
+
+      # TODO: Work out whether we really need the radio buttons here and hence
+      # whether we even need the `confirmation` attribute here.
+      if confirmed?
+        ConfirmOfferConditions.new(actor:, application_choice:, updated_conditions: false).save
```
Comment (stevehook) [inline app/forms/provider_interface/recruit_with_pending_conditions_form.rb:16]: Need to check the `updated_conditions` option here (doesn't look like it's implemented by the service).

Comment (stevehook) [reply app/forms/provider_interface/recruit_with_pending_conditions_form.rb:16]: Consider adding a new more specific service (with a different state machine event).

Comment (JR-G) [reply app/forms/provider_interface/recruit_with_pending_conditions_form.rb:16]: agree – could we use a separate service for this, thanks!

File: app/services/can_recruit_with_pending_conditions.rb [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+  end
+
+  def offer_has_no_unmet_non_ske_conditions?
+    application_choice.offer.conditions.none? do |condition|
+      !condition.is_a?(SkeCondition) && !condition.met?
+    end
+  end
+
+  def provider_is_scitt?
+    application_choice.provider&.provider_type&.to_s == SupportInterface::ProvidersFilter::SCITT
```
Comment (JR-G) [inline app/services/can_recruit_with_pending_conditions.rb:40]: Happy for this to be on the provider model. so we could just do `provider.scitt?`

Comment (stevehook) [reply app/services/can_recruit_with_pending_conditions.rb:40]: :+1: If it were an enum we'd get this for free I guess.

File: app/forms/provider_interface/recruit_with_pending_conditions_form.rb [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+
+    attr_accessor :actor, :application_choice, :confirmation
+
+    validates :confirmation, presence: true
+    validates :confirmation, inclusion: { in: %w[yes no] }
+
+    def save
+      return false unless valid?
+
+      if confirmed?
```
Comment (carlosmartinez) [inline app/forms/provider_interface/recruit_with_pending_conditions_form.rb:13]:
```suggestion     
      if valid? && confirmed?
```

File: app/services/can_recruit_with_pending_conditions.rb [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+
+  def initialize(application_choice:)
+    self.application_choice = application_choice
+  end
+
+  def call
+    feature_flag_enabled? &&
+      application_has_offer? &&
+      offer_has_pending_ske_conditions? &&
+      offer_has_no_unmet_non_ske_conditions? &&
```
Comment (carlosmartinez) [inline app/services/can_recruit_with_pending_conditions.rb:12]: this triple-negative name is a bit of a brainteaser. how about `all_non_ske_conditions_met?`

File: app/services/can_recruit_with_pending_conditions.rb [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+  end
+
+  def application_has_offer?
+    application_choice.offer.present?
+  end
+
+  def offer_has_pending_ske_conditions?
+    application_choice.offer.conditions.any? do |condition|
+      condition.is_a?(SkeCondition) && condition.pending?
+    end
```
Comment (carlosmartinez) [inline app/services/can_recruit_with_pending_conditions.rb:30]:
```suggestion
    application_choice.offer.ske_conditions.any?(&:pending?)
```

File: app/services/can_recruit_with_pending_conditions.rb [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+  end
+
+  def offer_has_no_unmet_non_ske_conditions?
+    application_choice.offer.conditions.none? do |condition|
+      !condition.is_a?(SkeCondition) && !condition.met?
+    end
```
Comment (carlosmartinez) [inline app/services/can_recruit_with_pending_conditions.rb:36]:
```suggestion
    non_ske_conditions = application_choice.offer.conditions - application_choice.offer.ske_conditions
    non_ske_conditions.all?(&:met?)
```

**Comment by carlosmartinez** [comment]
looks great @stevehook

File: app/services/can_recruit_with_pending_conditions.rb (hunk 4 above)
Comment (stevehook) [reply app/services/can_recruit_with_pending_conditions.rb:12]: fair point!

File: app/forms/provider_interface/recruit_with_pending_conditions_form.rb (hunk 3 above)
Comment (stevehook) [reply app/forms/provider_interface/recruit_with_pending_conditions_form.rb]: I think I still have to handle the _No_ radio button correctly (it's not the same end result as a validation fail).

**Comment by stevehook** [comment]
> **Database-level enum changes detected**
> 
> Please include a data migration for these attributes and values:
> 
> ```
> 4c4
> <   �[1m�[36mActiveRecord::SchemaMigration Pluck (1.9ms)�[0m  �[1m�[34mSELECT "schema_migrations"."version" FROM "schema_migrations" ORDER BY "schema_migrations"."version" ASC�[0m
> ---
> >   �[1m�[36mActiveRecord::SchemaMigration Pluck (2.1ms)�[0m  �[1m�[34mSELECT "schema_migrations"."version" FROM "schema_migrations" ORDER BY "schema_migrations"."version" ASC�[0m
> 5a6,8
> >   �[1m�[36mTRANSACTION (0.2ms)�[0m  �[1m�[35mBEGIN�[0m
> >   �[1m�[36mAudited::Audit Create (1.8ms)�[0m  �[1m�[32mINSERT INTO "audits" ("auditable_id", "auditable_type", "associated_id", "associated_type", "user_id", "user_type", "username", "action", "audited_changes", "version", "comment", "remote_address", "request_uuid", "created_at") VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11, $12, $13, $14) RETURNING "id"�[0m  [["auditable_id", 1], ["auditable_type", "SiteSetting"], ["associated_id", nil], ["associated_type", nil], ["user_id", nil], ["user_type", nil], ["username", nil], ["action", "create"], ["audited_changes", "{\"name\":\"cycle_schedule\",\"value\":null}"], ["version", 1], ["comment", nil], ["remote_address", nil], ["request_uuid", "59dddbc7-4763-44bd-81a1-b2dcbeb18400"], ["created_at", "2023-10-12 16:13:26.854505"]]
> >   �[1m�[36mTRANSACTION (0.5ms)�[0m  �[1m�[35mCOMMIT�[0m
> 8a12
> > <enum> Provider.provider_type: ["lead_school", "scitt", "university"]
> ```

Have checked production and the only values in use right now are `Provider.pluck(:provider_type).uniq
=> ["scitt", "university", "lead_school", nil]` so I don't think we have any data to migrate to the new enum.

**Review by carlosmartinez — APPROVED** [review]

**Review by JR-G — APPROVED** [review]

**Review by MaeveRoseveare — APPROVED** [review]

**Pull request merged**